### DISCUSSION TITLE
Signature of "method" incompatible with supertype "Class" [override]

### DISCUSSION THREAD

**Discussion by adam-grant-hendry** [python/typing#1219] [Q&A] [answered]
I am using `mypy==0.961` exclusively (not any other static type checker; e.g. `pyright`, `pyre`, `pytyped`, etc.)

I've imported `qtpy` with `PySide6` in `python==3.8.10` on Windows 10 x64-bit in VSCode 1.68.1. I have the following type annotation for `QtWidgets.QTabWidget.addTab`

```python
from __future__ import annotations

@overload
def addTab(
    self,
    __arg__2: str,
) -> int:
    ...

def addTab(
    self,
    __widget: PySide6.QtWidgets.QWidget,
    __label: str,
) -> int:
    ...    
```

`PySide6` is a binding of the C++ `Qt6` library using `shiboken6`. C++ does not natively support keyword arguments and `PySide6` has chosen to support only exporting optional arguments from `Qt6` as keyword arguments.

The [documentation for PySide6](https://doc.qt.io/qtforpython-6/PySide6/QtWidgets/QTabWidget.html#PySide6.QtWidgets.PySide6.QtWidgets.QTabWidget.addTab) seems to indicate my stubs are correct, but I could be wrong.

In my source, I have a class that subclasses `QTabWidget` and tries to overwrite the method:

```python
from qtpy import QtWidgets

class MyTabWidget(QtWidgets.QTabWidget):
    def addTab(
        self,
        widget: QtWidgets.QWidget,
        icon: QtGui.QIcon | QtGui.QPixmap | None,
        label: str,
    ) -> int:
        # Implementation here
```

However, `mypy` errors that the signatures are incompatible:

![image](https://user-images.githubusercontent.com/59346180/177402904-bbe7c3dd-dad3-40a9-bd62-4615a30f82e4.png)

I suspect the problem might be due to [Issue #11004](https://github.com/python/mypy/issues/11004), but I'm not sure.

Can anyone tell me if I am doing something wrong or if this is a bug I should `# type: ignore`?

**Comment by Akuli** [comment]
The mypy error is trying to say that `MyTabWidget().addTab(some_widget, "hello")` doesn't work. It would pass `"hello"` as `icon`, not as `label`. But there's an overload in `QTabWidget` that says it should work.

A simple solution would be to define your `addTab` like this:

```python3
def addTab(
    self,
    widget: QtWidgets.QWidget,
    icon: QtGui.QIcon | QtGui.QPixmap | str | None,
    label: str = "",
) -> int:
    if isinstance(icon, str):
        label = icon
        icon = None
    ...
```

It doesn't error in some cases where you would want an error though, such as `your_tab_widget.addTab(some_widget, "foo", "bar")`, but it works. You can improve on this a bit by copy/pasting the overloads from the pyqt stubs:

```python3
@overload
def addTab(
    self,
    __widget: PySide6.QtWidgets.QWidget,
    __label: str,
) -> int:
    ...    
@overload
def addTab(
    self,
    __widget: PySide6.QtWidgets.QWidget,
    __icon: QtGui.QIcon | QtGui.QPixmap | str | None,
    __label: str,
) -> int:
    ...    

def addTab(
    self,
    widget: QtWidgets.QWidget,
    icon: QtGui.QIcon | QtGui.QPixmap | str | None,
    label: str = "",
) -> int:
    if isinstance(icon, str):
        label = icon
        icon = None
    ...
```

In general, overriding overloaded methods isn't pretty. Are you sure that overriding this method is a good idea? Why not use "has-a" instead of "is-a" here, and just define your own method that calls `self.tab_widget.addTab(...)` or something like that?

**Reply by adam-grant-hendry** [reply]
@Akuli Thank you for your help!

> You can improve on this a bit by copy/pasting the overloads from the pyqt stubs

I checked

- [python-qt-tools/PyQt6-stubs](https://github.com/python-qt-tools/PyQt6-stubs/blob/main/PyQt6-stubs/QtWidgets.pyi)
- [python-qt-tools/PyQt5-stubs](https://github.com/python-qt-tools/PyQt5-stubs/blob/master/PyQt5-stubs/QtWidgets.pyi)

and didn't find the implementation you showed. Did you write this, or are you pulling from a different stubs repo?

> In general, overriding overloaded methods isn't pretty. Are you sure that overriding this method is a good idea?

That's a possibility...I need the widget to be substitutable wherever `QTabWidget` can be, so at a minimum I might need to make it subclas from `QWidget`, but it might be possible...

**Reply by Akuli** [reply]
> Did you write this, or are you pulling from a different stubs repo?

I wrote it based on the error message. It shows how the method is stubbed under the "Superclass:" part.

Apparently PySide6 comes with a bunch of `.pyi` files. To find out where they are, just `import PySide6` and `print(PySide6)`. `QTabWidget` is defined in `QtWidgets.pyi`, and you can find the same overloads there.

> I need the widget to be substitutable wherever QTabWidget can be

Why is this? If you are writing a library and the class is meant to be a drop-in replacement, I get it, and the best solution is to just copy/paste the overloads. If you are writing application code, just change whatever code is using the class.

**Reply by Akuli** [reply]
If you don't like copy/pasting, you could also use my `@copy_type` decorator: https://github.com/python/typing/issues/769#issuecomment-903760354

**Comment by adam-grant-hendry** [comment]
> Apparently PySide6 comes with a bunch of .pyi files.

Ya, I forgot you get the script `pyside6-genpyi` when you `pip install PySide6`:

```bash
$ pyside6-genpyi all
```

generates the stubs, but even those use:

```
@overload
def addTab(self, icon: Union[PySide6.QtGui.QIcon, PySide6.QtGui.QPixmap], text: str) -> int: ...
@overload
def addTab(self, text: str) -> int: ...
```

so it seems you're right and overloads are going to be a bit messy...

> If you are writing a library and the class is meant to be a drop-in replacement, I get it, and the best solution is to just copy/paste the overloads.

Bingo. It's a library. I'll definitely copy/paste the overloads.

**Reply by adam-grant-hendry** [reply]
@Akuli 

I discovered [Please add documentation about needing to re-define inherited overload signatures #10699](https://github.com/python/mypy/issues/10699) after your comment that I should copy/paste the overloads. I wasn't aware of this!

I actually tried the following:

```python
import typing

from qtpy import QtWidgets, QtGui

class MyTabWidget(QtWidgets.QTabWidget):
    @typing.overload
    def addTab(  # noqa: D102
        self,
        __widget: QtWidgets.QWidget,
        __arg__2: str,
    ) -> int:
        ...

    @typing.overload
    def addTab(  # noqa: D102
        self,
        __widget: QtWidgets.QWidget,
        __icon: QtGui.QIcon | QtGui.QPixmap | None,
        __label: str,
    ) -> int:
        ...

    def addTab(
        self,
        widget: QtWidgets.QWidget,
        icon: QtGui.QIcon | QtGui.QPixmap | str | None,
        label: str,
    ) -> int:
        # Implementation here...   

```

and I still get the error

```
Overloaded function implementation does not accept all possible arguments of signature 1 [misc]
```

![image](https://user-images.githubusercontent.com/59346180/177421465-affef22b-d399-4bb2-834a-4e641cc671d6.png)

**Comment by adam-grant-hendry** [comment, marked as answer]
I found the actual answer is this:

```python
class MyTabWidget(QtWidgets.QTabWidget):
    @typing.overload
    def addTab(
        self,
        widget: QtWidgets.QWidget,
        arg__2: str,
        /,
    ) -> int:  # noqa: D102  # Overloads don't get docstrings
        ...

    @typing.overload
    def addTab(
        self,
        widget: QtWidgets.QWidget,
        icon: QtGui.QIcon | QtGui.QPixmap,
        label: str,
        /,
    ) -> int:  # noqa: D102  # Overloads don't get docstrings
        ...

    def addTab(
        self,
        widget: QtWidgets.QWidget,
        icon: QtGui.QIcon | QtGui.QPixmap | str,
        label: str = '',
        /,
    ) -> int:
    # Implementation goes here...
```

For details, see [Overloaded function implementation does not accept all possible arguments of signature #13077](https://github.com/python/mypy/issues/13077)
